.github/workflows/main.yml

```diff
@@ -12,37 +12,52 @@ jobs:
         python-version: [3.7]
 
     steps:
-    - uses: actions/checkout@v1
-    - name: Set up Python ${{ matrix.python-version }}
-      uses: actions/setup-python@v1
-      with:
-        python-version: ${{ matrix.python-version }}
-
-    - name: Test formatting
-      run: |
-        pip install black
-        black --check .
-
-    - name: Install dependencies
-      run: |
-        python -m pip install --upgrade pip
-        pip install -r requirements.txt
-
-    - name: Test with pytest
-      run: |
-        pip install pytest
-        pytest
-
-    - name: Publish to Docker Registry
-      uses: elgohr/Publish-Docker-Github-Action@master
-      with:
-        name: simplelogin/app-ci
-        username: ${{ secrets.DOCKER_USERNAME }}
-        password: ${{ secrets.DOCKER_PASSWORD }}
-
-    - name: Send Telegram message
-      uses: appleboy/telegram-action@master
-      with:
-        to: ${{ secrets.TELEGRAM_TO }}
-        token: ${{ secrets.TELEGRAM_TOKEN }}
-        args: Docker image pushed on ${{ github.ref }}
+      - uses: actions/checkout@v2
+        
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+
+
+      - uses: actions/cache@v1
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements.txt') }}
+          restore-keys: |
+            ${{ runner.os }}-pip-
+
+      - name: Test formatting
+        run: |
+          pip install black
+          black --check .
```

Comment:
That will be a good idea with Nodejs since it can configure `dev` environment packages. 

However, I think `black` is currently cached. The hash of `requirement.txt` passed to the `action/cache` is just to flush the cache when there is some requirement changed. Even we didn't specify `black` in the requirement, it should be cached since it will still go to the cached folder.  

The only issue may be if there is no requirement upgraded but `black`, action will not cache that version of `black`. But this sounds not a big deal.